H2MLauncher.Core/Services/MatchmakingService.cs

```diff
@@ -0,0 +1,326 @@
+﻿using System.Net;
+
+using H2MLauncher.Core.Models;
+using H2MLauncher.Core.Settings;
+
+using Microsoft.AspNetCore.SignalR.Client;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace H2MLauncher.Core.Services
+{
+    public enum DequeueReason
+    {
+        Unknown,
+        UserLeave,
+        MaxJoinAttemptsReached,
+        JoinTimeout,
+        Disconnect,
+        Joined,
+        JoinFailed
+    }
+
+    public enum PlayerState
+    {
+        Disconnected,
+        Connected,
+        Queued,
+        Joining,
+        Joined
+    }
+
+    public sealed class MatchmakingService : IAsyncDisposable
+    {
+        private readonly H2MCommunicationService _h2MCommunicationService;
+
+        private readonly HubConnection _connection;
+        private readonly ILogger<MatchmakingService> _logger;
+        private readonly IGameCommunicationService _gameCommunicationService;
+        private readonly IOptionsMonitor<H2MLauncherSettings> _options;
+
+        private record QueuedServer(string Ip, int Port)
+        {
+            public bool HasSeenConnecting { get; set; }
+        }
+
+        private readonly List<QueuedServer> _queuedServers = [];
+        private bool _hasSeenConnecting = false;
+        private readonly Dictionary<(string ip, int port), string> _privatePasswords = [];
+
+        public int QueuePosition { get; private set; }
+        public int TotalPlayersInQueue { get; private set; }
+
+        private PlayerState _queueingState = PlayerState.Disconnected;
+        public PlayerState QueueingState
+        {
+            get => _queueingState;
+            set
+            {
+                if (_queueingState == value)
+                {
+                    return;
+                }
+
+                _logger.LogTrace("Set queueing state to {queueingState}", value);
+
+                PlayerState oldState = _queueingState;
+                _queueingState = value;
+
+                OnQueueingStateChanged(oldState, value);
+            }
+        }
+
+        public bool IsConnected => _connection.State is HubConnectionState.Connected;
+
+        public event Action<PlayerState>? QueueingStateChanged;
+        public event Action<int, int>? QueuePositionChanged;
+        public event Action<(string ip, int port)>? Joining;
+        public event Action<(string ip, int port)>? Joined;
+        public event Action<(string ip, int port)>? JoinFailed;
+
+        public MatchmakingService(ILogger<MatchmakingService> logger, H2MCommunicationService h2MCommunicationService,
+            IGameCommunicationService gameCommunicationService, IOptions<MatchmakingSettings> matchmakingSettings, IOptionsMonitor<H2MLauncherSettings> options)
```

Comment:
After 3 constructor parameters, I think it is nice to put them all on their own line.